KAFKA-18182: KIP-891 Connect Multiversion Support (Base PR with Plugin Loading Isolation Changes)

snehashisp opened the pull request:
The is one of a set of PRs for [KIP-891](https://cwiki.apache.org/confluence/display/KAFKA/KIP-891%3A+Running+multiple+versions+of+Connector+plugins). The list of total PRs given below all build one the previous one in the list. They can be reviewed individually, or if the complete set of changes is preferrable, please refer to the last PR. 

This is PR#1 and contains changes to connects plugin loading package. It contains the set of changes to load versioned plugins based on maven versioning as described in the KIP and exposes a set of public methods for use in the other components of this KIP. 

1. https://github.com/apache/kafka/pull/16984 - Plugins Loading Isolation Changes
2. https://github.com/apache/kafka/pull/17741 - Versioned Configs and Validation for Connector and Converters
3. https://github.com/apache/kafka/pull/17742 - Transformation and Predicate Support
4. https://github.com/apache/kafka/pull/17743 - Versioned Connector Creation 
5. https://github.com/apache/kafka/pull/17988 - Updates to status and metrics



### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

snehashisp committed:
Add multiversioning apis to isolation

snehashisp committed:
add method for header converter

snehashisp committed:
add getter for all plugins for a class

snehashisp committed:
Allow super to load class but do a version check

snehashisp committed:
validate the version when loaded from parent correctly

snehashisp committed:
bug in loadclass, should use fullname

snehashisp committed:
add available version to exception and minor changes

snehashisp committed:
Add default version getter

snehashisp committed:
add a null check

snehashisp committed:
Update log and comment

snehashisp committed:
Rename default version to latest version

snehashisp committed:
fix header converter version

snehashisp committed:
Use current classloader when version is not present

snehashisp committed:
Remove transformation and predicate getters

snehashisp committed:
add rawtype annotation

snehashisp committed:
Add a static loader swap method

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -122,11 +196,54 @@ protected Class<?> loadClass(String name, boolean resolve) throws ClassNotFoundE
+            log.trace("Retrieving loaded class '{}' from '{}'", name, pluginLoader);
+            plugin = pluginLoader.loadClass(fullName, resolve);
+        } else {
+            plugin = super.loadClass(fullName, resolve);
+            // if we are loading a plugin class from the parent classloader, we need to check if the version
+            // matches the range
+            String pluginVersion;
+            try (LoaderSwap classLoader = PluginScanner.withClassLoader(plugin.getClassLoader())) {
+                pluginVersion = PluginScanner.versionFor(plugin.getDeclaredConstructor().newInstance());
+            } catch (ReflectiveOperationException | LinkageError e) {
+                throw new VersionedPluginLoadingException(String.format(
+                        "Plugin %s was loaded with %s but failed to determine its version",
+                        name,
+                        plugin.getClassLoader()
+                ), e);
+            }
```
This information should be computed ahead-of-time by the PluginScanner, not re-computed at runtime each time the class is requested. The parent loader of the DelegatingClassLoader will always be considered "classpath" in PluginDesc<> objects, which also include the version.

snehashisp replied:
Is this part not really required? IIUC the plugins present in classpath are also scanned and part of the `pluginLoaders` map, and it should be found with the `pluginClassLoader` logic. `super.loadClass` should always throw a `ClassNotFoundException`. Otherwise, if there are edge cases where super.loadClass does find the class (maybe the plugin is part of the Bootstrap loader) then instantiation the class here is the only way to confirm version requirements are addressed.

gharris1727 replied:
> IIUC the plugins present in classpath are also scanned and part of the pluginLoaders map, and it should be found with the pluginClassLoader logic.

`pluginClassLoader()` returns `PluginClassLoader`, and has an instanceof check to filter out the classpath plugins.

> super.loadClass should always throw a ClassNotFoundException.

I don't know what you mean. If there's a copy of the requested plugin on the classpath, super.loadClass will find it, and it will already have been scanned and put in pluginLoaders with the location being `"classpath"`

snehashisp replied:
Ahh, I missed the instance of check in the `pluginClassLoader`. Yes in that case super.loadClass will load the plugin from classpath and we can avoid instantiation to do the version check. Will make the requested changes for this and the following comment.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +70,109 @@ public DelegatingClassLoader() {
+        String fullName = aliases.getOrDefault(connectorClassOrAlias, connectorClassOrAlias);
+        // if the plugin is not loaded via the plugin classloader, it might still be available in the parent delegating
+        // classloader, in order to check if the version satisfies the requirement we need to load the plugin class here
+        ClassLoader classLoader = loadVersionedPluginClass(fullName, range, false).getClassLoader();
```
There's a slight difference here: If the plugin is non-isolated, `DelegatingClassLoader#connectorLoader(String)` returns `this` (the delegating loader), but `DelegatingClassLoader#connectorLoader(String, VersionRange)` returns `parent` (the classpath).

For a similar reason, `connectorLoader(String)` doesn't throw `ClassNotFoundException`, it returns `this`.

snehashisp replied:
Similar query here (w.r.t my previous comment). Since the class path plugins are scanned and added to pluginLoaders dict, both should return the classpath loader. Its only if for some reason, a plugin is found that was not scanned we should get a different in the two methods. 

I also saw the other comment in the other PR. I have actually replaced the use of connector loader with fetching the connector and using `getClass().getClassLoader()` as that avoids two calls. Now in both cases it should return classpath loader for non-isolated plugins based on my understanding. This would mean that instantiation plugins with these classloaders will bypass the delegating loader, which does not look correct, but that seems to be the current process. LMK if I am missing something here. I can check if the class loader here is an instance of `PluginClassLoader` and return the delegating classloader here if I am incorrect in my assumption.

gharris1727 replied:
> but that seems to be the current process.

It's the process in your PR, but was not the process before. I'm saying you need to change this to return the DelegatingClassLoader.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +70,109 @@ public DelegatingClassLoader() {
+            : null;
+    }
+
+    PluginClassLoader pluginClassLoader(String name) {
```
I like this style of deduplication. Can you apply it to `connectorLoader` and `loadClass`/`loadVersionedPluginClass`?

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +70,109 @@ public DelegatingClassLoader() {
+            }
+
+            if (version != null) {
+                for (Map.Entry<PluginDesc<?>, ClassLoader> entry : loaders.entrySet()) {
```
Since this iterates forwards, when there are two PluginDesc objects with the same version, it will prefer the earlier one, which could be the classpath copy. We should prefer the later isolated one instead

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +70,109 @@ public DelegatingClassLoader() {
+
+            if (range.hasRestrictions()) {
+                List<ArtifactVersion> versions = loaders.keySet().stream().map(PluginDesc::encodedVersion).collect(Collectors.toList());
+                version = range.matchVersion(versions);
```
nit: you can avoid iterating over the loaders twice if you use VersionRange#containsVersion instead.

snehashisp replied:
Thanks, have altered the logic as per your suggestion, this should also address the other [comment](https://github.com/apache/kafka/pull/16984#discussion_r1868472074)on picking the latest plugin.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +70,109 @@ public DelegatingClassLoader() {
+
+        if (range != null) {
+
+            ArtifactVersion version = range.getRecommendedVersion();
```
Since we're not implementing soft requirements, i think getRecommendedVersion should always return null, and hasRestrictions always returns true. Rather than implementing this case which we don't expect to be used, can we avoid calling range.getRecommendedVersion entirely?

If a soft requirement somehow gets passed in here, it could either be a no-op (behaving like null) or throw an exception, up to you.

snehashisp replied:
Have opted to throw a exception here as we should not be getting a soft requirement here.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -240,6 +261,21 @@ public Runnable withClassLoader(ClassLoader classLoader, Runnable operation) {
         };
     }
 
+    public static LoaderSwap swapLoader(ClassLoader loader) {
```
I would prefer to keep this method (and the similar copy in PluginScanner) as instance methods, because they are simpler to mock during testing: #12817

snehashisp replied:
I made this static so that I can directly invoke it from the runtime classes while swapping loaders for converters, transformations. Making this non-static is an option but we should then move this logic to the runtime (or pass the plugins instance to the WorkerTasks which seems overkill). I will lead to a small code duplication but will aid in mocking and testing there.

gharris1727 replied:
So far the Worker has maintained the Plugins instance, and the WorkerTask has maintained the ClassLoader instance, allowing `plugins.swapLoader(task.loader())` to be called when necessary.

> pass the plugins instance to the WorkerTasks which seems overkill

I think this is the best way forward.

gharris1727 replied:
I realize now that there's a lot of classloaders flying around inside a WorkerTask, and loader() really doesn't cover them all.

Your calls to swapLoader in convertAndTransformRecord are actually new, and fixes an existing TCCL bug.
Fix it fully would require ClassLoader fields for each converter/predicate/transform/etc.

We can either leave it unaddressed, or throw in some lazy plugin.swapLoader(plugin.connectorLoader(pluginName, version)) calls...

What a mess. This is exactly why I wanted to get https://issues.apache.org/jira/browse/KAFKA-14670 done :(

gharris1727 replied:
Okay I thought about this some more.

The fact that this is static can be fixed later. But the fact that we're using plugin.getClass().getClassLoader() to get the ClassLoader is incorrect, and I think either needs a Plugins instance to call connectorLoader, or needs new ClassLoader fields anyway, so it may not be that much work to add the Plugins field as well.

The only thing we need to avoid is accidentally swapping to the AppClassLoader (parent of DelegatingClassLoader) instead of the DelegatingClassLoader itself, because that actually would make some classes less visible to plugins.

snehashisp replied:
Actually, there are classloader swaps required in transformations as well. I always wondered why separate plugins like connector and transformation, were running under the connectors class loader in workertasks. I guess up until now it would not cause issues as class loading would go via delegation to the correct class loader. But with versioned plugins the delegation would end up on the latest version of the plugin which can differ from the versioned plugin. That was my reasoning for introducing loader swaps under converters and transformations. The breath of places which required it led me to make it static for easy access. Also, I did not unnecessarily want to inject the entire plugin instance into those classes. 

I think we would want to avoid the lookup cost of finding the classloader from the plugins instance every time and then swapping the loader, so we would need to maintain a set of extra class loader variables for each instance in the worker tasks and transformations. I was thinking of exposing a `safeLoaderSwap` java function like below and pass that on to the worker tasks and transformations.
```
public Function<ClassLoader, LoaderSwap> safeLoaderSwapper() {
        return loader -> {
            if (!(loader instanceof PluginClassLoader)) {
                loader = delegatingLoader;
            }
            return withClassLoader(loader);
        };
    }
```
That way we don't have to keep track of each class loader. wdyt?

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -259,48 +299,91 @@ public Set<PluginDesc<SinkConnector>> sinkConnectors() {
         return scanResult.connectorClientConfigPolicies();
     }
 
+    private <T> Set<PluginDesc<T>> pluginsOfClass(String classNameOrAlias, Set<PluginDesc<T>> allPluginsOfType) {
```
This (and the methods calling it) are a lot of additional API surface area for the Plugins class, which is already quite large. What benefits does this "lookup versions of plugins by alias" functionality provide to compensate for the extra surface area?

1. Populate a map with null values in `AbstractHerder#getConnector`
2. Generate a list of valid versions in Recommender instances (e.g. ConnectorPluginVersionRecommender)

I think (1) is very low value, and could be avoided with a null/contains check elsewhere, or could be performed for all plugins at once rather than lazily per-alias.
(2) is a valid use-case, but could better be located within a common PluginVersionRecommender class, and utilize the existing Plugins API surface.

snehashisp replied:
I thought about it initially, and while this can be done in other modules, it was simpler to add this capability here rather than having to compute it in two separate places in the AbstractHerder and Recommenders, especially because they don't have access to the alias dictionary which would require us to expose it from the Plugins module and provide appropriate logic to handle aliases. IMO the ownership of plugins description and lookups should be encapsulated within the Plugins module itself. I agree with this increasing the API footprint, but I seems like a reasonable compromise.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -259,48 +299,91 @@ public Set<PluginDesc<SinkConnector>> sinkConnectors() {
-            );
+            klass = range == null ?
+                pluginClass(delegatingLoader, connectorClassOrAlias, Connector.class):
+                pluginClass(delegatingLoader, connectorClassOrAlias, Connector.class, range);
```
The 4-argument `pluginClass` already handles null range, can you call that unconditionally?

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -350,54 +437,49 @@ public Task newTask(Class<? extends Task> taskClass) {
+    }
+
+    /**
+     * Used to get a versioned converter. It will always try and get the converter from the set of plugin classloaders.
+     *
+     * @param config              the configuration containing the {@link Converter}'s configuration; may not be null
+     * @param classPropertyName   the name of the property that contains the name of the {@link Converter} class; may not be null
+     * @param versionPropertyName the name of the property that contains the version of the {@link Converter} class; may not be null
+     * @return the instantiated and configured {@link Converter}; null if the configuration did not define the specified property
+     * @throws ConnectException if the {@link Converter} implementation class could not be found,
+     * @throws VersionedPluginLoadingException if the version requested is not found
+     */
+    public Converter newConverter(AbstractConfig config, String classPropertyName, String versionPropertyName) {
+        ClassLoaderUsage classLoader = config.getString(versionPropertyName) == null ? ClassLoaderUsage.CURRENT_CLASSLOADER: ClassLoaderUsage.PLUGINS;
```
Is this top comment misleading? The current classloader is used if no version is specified.

snehashisp replied:
Yes, its incorrect, should be only when the version is specified.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -434,50 +516,93 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
+        String basePluginClassName = basePluginClass.getSimpleName();
         switch (classLoaderUsage) {
             case CURRENT_CLASSLOADER:
-                if (!config.originals().containsKey(classPropertyName)) {
-                    // This connector configuration does not define the header converter via the specified property name
-                    return null;
-                }
```
Odd, the newHeaderConverter didn't perform this containsKey check when PLUGINS was specified, when the configuration is coming from the worker config.

But since `header.converter` has a default, it was always valid to call getClass. Now the `getHeaderConverter` method performs this check unconditionally, does this mean that Worker#startTask will sometimes not instantiate a header converter?

snehashisp replied:
Good point, I think if the header converter is not defined in both connector config and worker config it will not be initialized.

TBH, this initialization mechanism always seemed a bit brittle. Class loader usage seems to be strongly tied to whether we use the connector config or the worker config. Although we don't do it, there is nothing stopping us from passing a WorkerConfig with current class loader or vice versa. Also, CURRENT_CLASSLOADER seems to indicate that we are creating the plugin with the loader of the current thread. This is not exactly the case as its actually the loader in use when the AbstractConfig was initialized (which internally does Utils.loadclass for class type configs). This can be addressed separately but for now I have preserved the existing behavior to reduce potential edge cases.

gharris1727 replied:
> Although we don't do it, there is nothing stopping us from passing a WorkerConfig with current class loader or vice versa. 

Yeah we don't do that because it doesn't make much sense. CURRENT_CLASSLOADER implies that there's a particular PluginClassLoader which is given precedence. The WorkerConfig doesn't have a preferred PluginClassLoader, so it only ever makes sense to use PLUGINS.

Although CURRENT_CLASSLOADER when the TCCL is the DelegatingClassLoader should be equivalent to PLUGINS as far as I understand...

I agree that we don't have to fully clean-up the ClassLoaderUsage stuff immediately, just enough to keep compatible behaviors.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -350,54 +437,49 @@ public Task newTask(Class<? extends Task> taskClass) {
             // This configuration does not define the converter via the specified property name
             return null;
         }
-
```
I think it was a good idea to deduplicate this switch-case. There's another copy in newConfigProvider that we can get rid of too, and just have it use a `null` `VersionRange`.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -434,50 +516,93 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
             plugin = newPlugin(klass);
-            plugin.configure(converterConfig);
+            DefaultArtifactVersion pluginVersion = new DefaultArtifactVersion(PluginScanner.versionFor(plugin));
+            if (range != null && range.hasRestrictions() && !range.containsVersion(pluginVersion)) {
+                // this can happen if the current class loader is used
+                // if there are version restrictions then this should be captured, and we should load using the plugin class loader
+                if (classLoaderUsage == ClassLoaderUsage.CURRENT_CLASSLOADER) {
+                    return getVersionedPlugin(config, classPropertyName, versionPropertyName, basePluginClass, ClassLoaderUsage.PLUGINS, availablePlugins);
+                }
+            }
```
Can we assert `range == null || classLoaderUsage == PLUGINS` earlier, to ensure that we never get into this situation where we loaded a class of an invalid version?

I think having this depend on range.containsVersion(pluginVersion) could lead to some strange behavior in the following situation:
* key.converter.version = [1.0, 2.0)
* Converter v1.0 is on the classpath
* Converter v1.2 is in a different plugin
* Converter v1.1 is in the local plugin

If you don't have 1.1 installed, 1.2 is selected. If you later install 1.1 (an older version), you get rolled back to 1.1.
But if you had specified [1.2] it wouldn't get rolled back.

snehashisp replied:
Yes, the situation you mentioned is exactly the reason why the public methods to create a converter with a version provided we don't have the option to pass in a class loader and defaults to using the plugins loader, otherwise it can potentially shadow a higher version of a plugin requirement. This code path should ideally never be executed. Will remove this and add the assertion suggestion.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -16,17 +16,18 @@
-import java.util.Map;
-import java.util.SortedMap;
-import java.util.TreeMap;
+import java.util.*;
```
nit: no star imports

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -350,54 +437,49 @@ public Task newTask(Class<? extends Task> taskClass) {
+        return getConverter(config, classPropertyName, versionPropertyName, classLoader);
+    }
+
+    private Converter getConverter(AbstractConfig config, String classPropertyName, String versionPropertyName, ClassLoaderUsage classLoaderUsage) {
```
nit: Don't use the `get` prefix, I don't think it makes sense when the converter is being created. `new` is more appropriate here.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginLoadingException.java:
```diff
@@ -0,0 +1,32 @@
+
+import java.util.List;
+
+public class VersionedPluginLoadingException extends ConnectException {
```
Does this make more sense as a ConfigException?

snehashisp replied:
I'm okay with making it an extension of `ConfigException`. But let's keep it as a separate class type as it makes the errors more specific and helps in debugging. It also provides a set of available versions which is useful in other modules which capture this exception.

gharris1727 replied:
Yeah I was imagining `extends ConfigException`.

gharris1727 reviewed (commented):
Thanks so much for the PR @snehashisp.

I apologize for the lateness of my review. I'm planning on reviewing this daily until the feature freeze on the 11th.

snehashisp commented:
Thanks for the reviews @gharris1727. Will take a look soon and try and give my full attention till 11th. What does the feature freeze encompass, does it include all unit testing and ITs or can it come later as part of stabilization? I believe we can make it if it does not include the latter, otherwise we will need to target the following release.

gharris1727 commented:
> What does the feature freeze encompass, does it include all unit testing and ITs or can it come later as part of stabilization?

Earlier is always better, but we can prioritize the changes (with manual testing) before feature freeze, and unit/integration tests after feature freeze.

> I'll try my best, but 11th seems a bit ambitious given the scope of work here.

I'll do my best to support you, please let me know if you see anything that I can help with, or if we need to involve someone else.

snehashisp committed:
Resolve comments on 1st review

snehashisp committed:
Remove extra code

snehashisp committed:
Add plugin version utils for parsing versions

snehashisp committed:
Should return loader

snehashisp committed:
fix incorrect classloader equality check

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -1,132 +1,253 @@
-        return pluginLoaders;
-    }
-}
+/*
```
nit: is there a big whitespace change in this class? The diff looks huge.

snehashisp replied:
I'm not sure why this is the case tbh. Maybe just the magnitude of changes in the class is making the diff on github a bit disproportional 🤔

snehashisp committed:
fix whitespace issue

snehashisp committed:
Update version loading logic and add return delegating loader

snehashisp committed:
Combine loader logic

snehashisp committed:
Don't expose some plugins methods

snehashisp committed:
add safe swap loader instead of static swaploader

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +75,117 @@ public DelegatingClassLoader() {
+                }
+            }
+
+            if (version == null) {
```
nit: If you change this to loader == null you an eliminate the `version` variable entirely.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +75,117 @@ public DelegatingClassLoader() {
+            ClassLoader loader = null;
+            for (Map.Entry<PluginDesc<?>, ClassLoader> entry : loaders.entrySet()) {
+                // the entries should be in sorted order of versions so this should end up picking the latest version which matches the range
+                if (range.containsVersion(entry.getKey().encodedVersion()) && entry.getValue() instanceof PluginClassLoader) {
```
nit: this `instanceof` check can be removed because it duplicates the check in the caller.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -69,36 +75,117 @@ public DelegatingClassLoader() {
+        String fullName = aliases.getOrDefault(classOrAlias, classOrAlias);
+        ClassLoader classLoader = range == null ? pluginClassLoader(fullName) :
+                // load the plugin class when version is provided which will validate the correct version is loaded
+                loadVersionedPluginClass(fullName, range, false).getClassLoader();
```
This logic is incorrect, and I think you can just call `pluginClassLoader(fullName, range)` to let the null check internally handle both cases.

This will also remove the ClassNotFoundException and the need to catch it higher up.

snehashisp replied:
So the reason why I have this check in place is because the classloader instantiation for versioned plugins should ideally make sure that the correct one is returned. I think returning delegating loader would be incorrect (without the validation) since we are not checking if there is a different version of the plugin present in class path (or multiple versions). I think if the actual plugin instantiation is done appropriately (as opposed to something like a `Utils.newInstance` after swapping to the returned classloader) we should be fine.  LMK you thoughts on this.

gharris1727 replied:
> So the reason why I have this check in place is because the classloader instantiation for versioned plugins should ideally make sure that the correct one is returned

I'm not really sure I understand this justification.
1. If code is calling `#loader(String, VersionRange)` to get a ClassLoader, it's going to be to interact with a plugin with a specific version (instantiate, call methods, etc).
2. In order to get a plugin class or instance of a specific version, the caller will need to use `#loadVersionedPluginClass(String, VersionRange, boolean)` which will perform the verification.
3. If the other verification wasn't sufficient, then they must have used a different VersionRange.

So i think this exception would only get thrown if the caller made a mistake and used two different VersionRange objects (or a null one).
I do want to make this sort of mistake less likely, but I think throwing an exception is not the way to do that.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -112,21 +199,70 @@ public void installDiscoveredPlugins(PluginScanResult scanResult) {
+                                + "only one version is allowed in class path for loading a plugin with version range",
+                        fullName
+                ));
+            } else {
```
There's also the case of classpathPlugins.isEmpty(), in case loadClass returns a valid class, but it wasn't discovered by scanning, like if the class is abstract or not migrated.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -112,21 +199,70 @@ public void installDiscoveredPlugins(PluginScanResult scanResult) {
+                            range
+                    ), Collections.singletonList(pluginVersion));
+                }
+            }
```
nit: All of the verification steps after the range == null check can be moved to a separate method, maybe `ArtifactVersion classpathPluginVersion(String)` or `boolean classpathPluginVersionInRange(String, VersionRange)`.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/util/PluginVersionUtils.java:
```diff
@@ -0,0 +1,30 @@
+    private static Plugins plugins = null;
+
+    public static void setPlugins(Plugins plugins) {
+        PluginVersionUtils.plugins = plugins;
```
Did this leak in from the other PR?

snehashisp replied:
yes, this is now unnecessary will remove

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -427,99 +503,121 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
                 // Attempt to load first with the current classloader, and plugins as a fallback.
-                klass = pluginClassFromConfig(config, classPropertyName, ConfigProvider.class, scanResult.configProviders());
+                // Note: we can't use config.getConfiguredInstance because Converter doesn't implement Configurable, and even if it did
+                // we have to remove the property prefixes before calling config(...) and we still always want to call Converter.config.
```
nit: stale comment

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -427,99 +503,121 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
         if (klass == null) {
-            throw new ConnectException("Unable to initialize the ConfigProvider specified in '" + classPropertyName + "'");
+            throw new ConnectException("Unable to initialize the '" + basePluginClassName
+                    + "' specified in '" + classPropertyName + "'");
```
nit: no single quotes around basePluginClassName.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/VersionedPluginLoadingException.java:
```diff
@@ -0,0 +1,49 @@
+    }
+
+    public VersionedPluginLoadingException(String message, Throwable cause) {
+        super(message, cause);
+    }
+
+    public VersionedPluginLoadingException(String message, Throwable cause, List<String> availableVersions) {
+        super(message, cause);
```
Hmm this is not a helpful constructor, it doesn't set the `Throwable cause`.
Since these constructors appear unused, and VersionedPluginLoadingExceptions are from invalid constraints and not other operations failing, WDYT about removing these constructors?

snehashisp replied:
Yes, we can remove these constructors.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -427,99 +510,121 @@ public Converter newInternalConverter(boolean isKey, String className, Map<Strin
+            try {
+                range = PluginVersionUtils.connectorVersionRequirement(version);
+            } catch (InvalidVersionSpecificationException e) {
+                throw new ConnectException(String.format("Invalid version range for %s: %s %s", classPropertyName, version, e));
```
nit: move the exception to the `cause` argument.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/PluginScanner.java:
```diff
@@ -222,7 +222,7 @@ protected static String reflectiveErrorDescription(Throwable t) {
     }
 
-    protected LoaderSwap withClassLoader(ClassLoader loader) {
+    protected static LoaderSwap withClassLoader(ClassLoader loader) {
```
This can be nonstatic now that DelegatingLoader isn't using it.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/DelegatingClassLoader.java:
```diff
@@ -112,21 +199,70 @@ public void installDiscoveredPlugins(PluginScanResult scanResult) {
+        return plugin;
     }
 
+
+
```
nit: extra lines

snehashisp committed:
Resolve comments

snehashisp commented:
Thanks for the reviews on this PR @gharris1727. I think I have addressed most of the comments here and made the changes to the other PRs based on your suggestions. Please review the other PRs as well. Apologies if its a bit hard since all the changes build on top of the previous ones. I find that looking at the [diff](https://github.com/snehashisp/kafka/compare/multiversioning...snehashisp:kafka:mvn-validation?expand=1 )via compare makes it easier.

gharris1727 commented on connect/runtime/src/main/java/org/apache/kafka/connect/runtime/isolation/Plugins.java:
```diff
@@ -168,11 +174,21 @@ protected static <U> Class<? extends U> pluginClass(
+            Class<U> pluginClass,
+            VersionRange range
+    ) throws VersionedPluginLoadingException, ClassNotFoundException {
+        Class<?> klass = range != null ?
+            loader.loadVersionedPluginClass(classOrAlias, range, false) : loader.loadClass(classOrAlias, false);
```
This doesn't need to check if range is null.

gharris1727 reviewed (commented):
Please work on getting this to compile and tests to pass.

snehashisp committed:
resolve comments and fix some checkstyle

snehashisp committed:
fix more checkstyle errors

snehashisp committed:
Merge remote-tracking branch 'origin' into multiversioning

snehashisp committed:
Remove classpath helper

snehashisp committed:
fix tests

snehashisp commented:
> Please work on getting this to compile and tests to pass.

A couple of minor adjustments were needed, otherwise the runtime tests have all passed locally. Hoping that it will go through on the CI 🤞. I accidentally pushed in some changes to rest of the files while merging latest trunk (git merge and push from native windoes seems to add a carriage return to the files 😢).  If possible can you please remove the additional tags.

gharris1727 reviewed (approved):
Thanks so much @snehashisp !
